jsonlreader.go

```diff
@@ -0,0 +1,104 @@
+package giashard
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"io"
+	"log"
+	"os"
+
+	"github.com/klauspost/compress/zstd"
+)
+
+// was schema, now hardcoded
+type JsonlRecord struct {
+	Url  string `json:"u"`
+	Text string `json:"text"`
+}
+
+// support reading a zstandard-zipped JSONL file and sending lines to channel (from giashard/LineReader)
+// f: os.File value, z: zstd.Decoder value, fatal: indicator if read errors be fatal
+type JsonlReader struct {
+	f     io.ReadCloser
+	z     io.ReadCloser
+	fatal bool
+}
+
+func NewJsonlReader(filename string) (r *JsonlReader, err error) {
+	f, err := os.Open(filename)
+	if err != nil {
+		return
+	}
+	d, err := zstd.NewReader(f)
+	if err != nil {
+		return
+	}
+	z := d.IOReadCloser() // to match LineReader
+	r = &JsonlReader{f, z, true}
+	return
+}
+
+// should read errors be fatal (and abort the program with log.Fatalf)
+func (r *JsonlReader) Fatal(flag bool) {
+	r.fatal = flag
+}
+
+// close the underlying files, ta3ban
+func (r *JsonlReader) Close() (err error) {
+	if e := r.z.Close(); e != nil {
+		err = e
+	}
+	if e := r.f.Close(); e != nil {
+		err = e
+	}
+	return
+}
+
+// send records read from file to channel (replaces Lines())
+func (r *JsonlReader) Records() (ch chan JsonlRecord) {
+	ch = make(chan JsonlRecord)
+	decoder := json.NewDecoder(r.z)
+	go func() {
+		for decoder.More() {
+			var record JsonlRecord
+			if err := decoder.Decode(&record); err != nil {
+				if r.fatal {
+					log.Fatalf("Error decoding record: %v", err)
+				} else {
+					log.Printf("Error decoding record: %v", err)
+				}
```

Comment:
Do you want to support non-fatal errors? Can you actually recover from say syntax errors? Since you're using a hard-coded `true` for `fatal` anyway, I'd say keep code simple and just make that the only implementation (I know Graeme might want to have a word with me about this mindset hehe)

Reply:
I'll leave it, don't want to interfere with ww's grand plan 🙏